### PR TITLE
Backend tests

### PR DESCRIPTION
**Pull request by codingpop**

### CODE REVIEW

File: server/controllers/addUsers.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import express from 'express';
```
Comment (noordean) [inline server/controllers/addUsers.js:1]: An empty line is required after importing npm modules, before importing your files

File: server/controllers/addUsers.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            response.status(404).json({
+              message: "Users don't exist",
+            });
+          } else {
+            postIt.findGroup(request.params.groupId)
+              .then((groupData) => {
+                groupData.addUsers(users)
+                  .then((newMembers) => {
+                    if (!newMembers.length) {
+                      response.status(406).json({
```
Comment (noordean) [inline server/controllers/addUsers.js:34]: This should be conflict(409). User already added

File: server/controllers/createGroup.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Creates a new group
+ * Rejects request if the user is unauthenticated
+ * Refuses to create group if group name is empty
+ */
+createGroup.post('/groups', verifyToken, (request, response) => {
+  const { name, description } = request.body;
+
+  if (!name || !description) {
+    response.status(400).json({
```
Comment (noordean) [inline server/controllers/createGroup.js:17]: Try and handle more potential errors here, such as white-space. It's possible a name or description containing only white-space would pass through this endpoint, and i can't see any middleware you'v put in place

File: server/controllers/createGroup.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        group.addUser(request.user.userId)
+          .then(() => {
+            response.status(201).json({
+              group,
+            });
+          });
+      })
+      .catch((error) => {
+        if (error.name === 'SequelizeUniqueConstraintError') {
+          response.status(406).json({
```
Comment (noordean) [inline server/controllers/createGroup.js:32]: Here again, it should be 409

File: server/controllers/getGroups.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+const getGroups = express.Router();
+const postIt = new PostIt();
+
+/**
+ * Gets all the groups to which a user belongs
+ */
+getGroups.get('/groups', verifyToken, (request, response) => {
+  postIt.findUser(request.user.userId)
+    .then((userData) => {
```
Comment (noordean) [inline server/controllers/getGroups.js:13]: What if the supplied userId is not in the database?

File: server/controllers/getGroups.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          userData.getGroups({ limit, offset, order: [['createdAt', 'DESC']] })
+            .then((userGroups) => {
+              const refinedGroups = userGroups.map(group => (
+                {
+                  groupId: group.groupId,
+                  name: group.name,
+                  description: group.description,
+                }
+              ));
+              response.json({
```
Comment (noordean) [inline server/controllers/getGroups.js:28]: You didn't put status code here

File: server/controllers/getMembers.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        groupData.getUsers()
+          .then((users) => {
+            const refinedUsers = users.map(user => (
+              {
+                userId: user.userId,
+                userName: user.userName,
+                phone: user.phone,
+              }
+            ));
+            response.json({
```
Comment (noordean) [inline server/controllers/getMembers.js:28]: Here also. No status code

File: server/controllers/getMembers.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Gets all the members in a group
+ */
+getMembers.get(
+  '/groups/:groupId/users',
+  verifyToken,
+  verifyMembership,
+  (request, response) => {
+    postIt.findGroup(request.params.groupId)
```
Comment (noordean) [inline server/controllers/getMembers.js:17]: What if this group has no member ? I can't see where you sent 'not found'. Kindly look into this

File: server/controllers/postMessage.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  verifyToken,
+  verifyMembership,
+  (request, response) => {
+    const { priority, body } = request.body;
+    if (!request.body.body) {
+      response.status(400).json({
+        message: 'Message body is required'
+      });
+    } else {
+      postIt.postMessage(
```
Comment (noordean) [inline server/controllers/postMessage.js:23]: Try and handle more errors here. What if there's no group with the supplied groupId?

File: server/src/PostIt.js [hunk 10]
```diff
@@ -67,77 +110,46 @@ class PostIt {
...
+      author,
+      userName,
+      body,
+      priority,
+    });
   }
 
   /**
    * Fetches all messages in a particular group
    * @param {string} groupId - target group
```
Comment (noordean) [inline server/src/PostIt.js:145]: I think your groupId is an integer, and not string

File: test/apiTests.js [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+
+const expect = chai.expect;
+
+describe('Given an API request', () => {
+  before(async () => {
+    await database.connection.sync({ force: true });
+  });
+
+  describe('When the request contains appropriate user registration data', () => {
+    it('Then the user is registered', async () => {
```
Comment (noordean) [inline test/apiTests.js:17]: Try and improve your test descriptions....something like it('should ....') would be better

File: test/apiTests.js [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+
+    it('Then the user is registered', async () => {
+      await request(app)
+        .post('/api/v1/users/signup')
+        .set('Accept', 'application/x-www-form-urlencoded')
+        .send(testData.newUser3)
+        .expect(201)
+        .then((response) => {
+          expect(response.body).to.be.an('object');
+          expect(response.body.message).to.equal('Registration successful');
```
Comment (noordean) [inline test/apiTests.js:49]: Having tested that the response body is an object and that the object contains a message property with value 'Registration successful'. It's important you test for the user object in the res body

**Review by noordean — COMMENTED** [review]
You'v done a great job. Below are few areas that need more of your attention:
- error handling in your middlewares or controllers, there are more potential errors to be handled
- consider improving your test description
- assert for more test cases. Do check if the parameters in the req body contain the supplied value.

File: server/controllers/getMembers.js (hunk 8 above)
Comment (codingpop) [reply server/controllers/getMembers.js:17]: My code returns an empty array, which I also need on the frontend

File: server/controllers/createGroup.js (hunk 3 above)
Comment (codingpop) [reply server/controllers/createGroup.js:17]: A lot of the errors are handled in the middleware `verifyMembership.js`. Check it out

File: server/controllers/postMessage.js (hunk 9 above)
Comment (codingpop) [reply server/controllers/postMessage.js:23]: I handled enough errors. Pls check `verifyMembership.js`

File: server/controllers/createGroup.js (hunk 3 above)
Comment (noordean) [reply server/controllers/createGroup.js:17]: check createGroup.post().....the only middleware there is verifyToken

**Pull request closed without merging**